docs/proposals/train_api_proposal.md

```diff
@@ -0,0 +1,154 @@
+
+**<h3>Background</h3>**
+
+Currently, there are two flows for data scientists to start using Training operator for their distributed training needs. 
+
+**<h4>Traditional method</h4>**
+
+1. The end-user has to write a custom script that includes data fetching, model loading, and a training loop using pytorch libraries.
+2. Create an image and push it to a registry.
+3. Using the Kubeflow train sdk to create worker spec and master spec which uses the above image and using these two specs create a pytorch job.
+
+This method requires infrastructure knowledge including image creation, registry access and training job spec. This makes it difficult for data scientists to purely focus at the application layer.
+
+**<h4>Higher level SDK</h4>**
+
+To provide a better user experience, a new higher level SDK was added in[ https://github.com/kubeflow/training-operator/pull/1659](https://github.com/kubeflow/training-operator/pull/1659) which avoided the need of image creation. Instead of creating the image, we can directly pass the custom script to train_func parameter and final pytorch job is created
+
+```python
+training_client.create_job(
+   name=pytorchjob_name,
+   train_func=train_function,
+   num_worker_replicas=3, # How many PyTorch Workers will be created.
+)
+```
+
+The above method reduces the time taken but still the end user has to write the train_function method.
+
+In many cases like in LLM, training core code template is a standard and it doesn’t require any custom modification. For faster experimentation of such models, an easier api should exist to train/finetune models without writing custom Python code. HuggingFace Trainer class is a good example but it is not easy to set it up in a Kubernetes environment. Training operator can handle this complexity and provide a better abstracted layer to the user.
+
+**<h3>New Train API proposal:</h3>**
+
+```python
+From kubeflow.training import trainingClient
+# Arguments related to the model provider including credentials
+model_args = modelProviderClass()
+# Arguments related to the dataset provider including credentials
+dataset_args = datasetProviderClass()
+# Arguments related to the trainer code
+parameters = {key:value pairs}
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='provider://repo/model_name', 
+   dataset= 'provider://train_dataset_path',
+   eval_dataset = "provider://eval_dataset_path"
+   model_args, 
+   dataset_args, 
+   parameters
+)
+```
+
+Example: 
+
+```python
+@dataclass
+class HuggingFace:
+    access_token = field()
+    peft_config = field()
+    transformerClass = field()
+
+@dataclass
+class S3:
+    access_token = field()
+    region = field()
+ 
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
```

Comment:
Should these parameters be specific to PyTorch or be abstract ?
As @kuizhiqing and @tenzen-y mentioned above: https://github.com/kubeflow/training-operator/pull/1945#issuecomment-1821168280, we might support MPI and Deepspeed in the future via this API.
Maybe we should define these params as follows to be agnostic to the underlying framework:
```py
num_workers or num_nodes
resources_per_worker or resources_per_node

For example:
num_workers = 5,
resources_per_worker = {"gpu": "1"}
```

Like we did in Katib: https://github.com/kubeflow/katib/blob/master/sdk/python/v1beta1/kubeflow/katib/api/katib_client.py#L153.
That will give us more flexibility in the future to support other resource (e.g. TPUs).

WDYT @deepanker13 @johnugeorge @tenzen-y @kuizhiqing @droctothorpe ?

Reply:
Yes, I think you are right we should rename it.

Reply:
`nprocPerNode` in [PytorchJob](https://github.com/kubeflow/training-operator/blob/master/pkg/apis/kubeflow.org/v1/pytorch_types.go#L95) and `slotsPerWorker` in [MPIJob](https://github.com/kubeflow/mpi-operator/blob/master/pkg/apis/kubeflow/v2beta1/types.go#L155) are actually related to the same concept internally. 

However， unify the name while also accommodating customary usage is a tough work.

Specifically, I prefer `worker` other than `node` here.

Reply:
Or, it might be better to name it `process_per_node`. It will be recognized `nprocPerNode` in `PyTorchJob`, `slotsPerWorker` in `MPIJob`.

Reply:
I suggested to name it as resources_per_worker since we can configure other resources later (e.g. CPU, TPU, Memory for more advance users). WDYT @tenzen-y ?

Reply:
> I suggested to name it as `resources_per_worker` since we can configure other resources later (e.g. CPU, TPU, Memory for more advance users). WDYT @tenzen-y ?

I imagined that `nprocs_per_node` means how many processes we launch in a single Pod, right?
So, I thought that `process` is better since maybe we can not specify how many amount of resources (CPU, GPU, etc) we use the resources per process in a single Pod.

Is it correct?

Reply:
How do we configure Number of Processes for the training script ?
Are we doing this via `torchrun` command today?

Also, how we can assign GPU number/slice to the training process ? Is it user responsibility ?

Reply:
> How do we configure Number of Processes for the training script ?
Are we doing this via torchrun command today?

@andreyvelich I imagined that we could set it via `nproc_per_node`. In other frameworks like Tensorflow, I'm not sure that it is possible.

> Also, how we can assign device ID (GPU or CPU) to the training process inside the Kubernetes container ?
Is it user responsibility ?

I think it is the responsibility of users since assigning way depends on training implementations.

Reply:
> For eg:
> num_workers = 5,
> resources_per_worker = {"gpu": "2", "cpu":8, "memory": "16Gi"}
> 
> sets the job.Spec.PyTorchReplicaSpecs["Worker"].Replicas to 5 with each WorkerReplica PodTemplateSpec set to have 2 GPU, 8 CPU and 16G memory. In addition, Spec.NprocPerNode will be set to 2(equal to the number of GPUs per node)

@johnugeorge I don't think it is possible. AFAIK, although we can set how many processes we launch in a single Pod, we can not set how many resources (gpu/cpu/memory) we assign per process in a single pod.

Reply:
@tenzen-y Looking [at the PyTorch docs](https://pytorch.org/docs/stable/distributed.html#launch-utility), it seems like when we use GPU, each proc consumes a single GPU, but in case of CPU training we can't define it. It looks like `nproc_per_node` is very specific parameter to `torchrun`.
Maybe initially, we should introduce these 3 parameters:
```
num_workers
resource_per_worker
num_procs_per_worker or nom_slots_per_worker
```
That should cover the `torchrun` and `mpi`.
WDYT @johnugeorge @tenzen-y @kuizhiqing ?

Reply:
Sounds good. Had a sync with @tenzen-y as well
We will use `num_workers`, `resource_per_worker` and `num_procs_per_worker`

For eg:
`num_workers` = 5,
`num_procs_per_worker` = 1
`resources_per_worker` = {"gpu": "2", "cpu":8, "memory": "16Gi"}

sets the `job.Spec.PyTorchReplicaSpecs["Worker"].Replicas` to 5 with each `WorkerReplica` PodTemplateSpec set to have 2 GPU, 8 CPU and 16G memory. In addition, `Spec.NprocPerNode` will be set to 1